core/src/main/scala/quasar/destination/s3/impl/DefaultUpload.scala

```diff
@@ -0,0 +1,128 @@
+
+import cats.effect.{Concurrent, ExitCase}
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import fs2.Stream
+import monix.catnap.syntax._
+import software.amazon.awssdk.services.s3.S3AsyncClient
+import software.amazon.awssdk.core.async.AsyncRequestBody
+import software.amazon.awssdk.services.s3.S3AsyncClient
+import software.amazon.awssdk.services.s3.model.{
+  AbortMultipartUploadRequest,
+  AbortMultipartUploadResponse,
+  CompleteMultipartUploadRequest,
+  CompleteMultipartUploadResponse,
+  CompletedMultipartUpload,
+  CompletedPart,
+  CreateMultipartUploadRequest,
+  CreateMultipartUploadResponse,
+  UploadPartRequest
+}
+
+final case class DefaultUpload[F[_]: Concurrent](client: S3AsyncClient, partSize: Int) extends Upload[F] {
+  def upload(bytes: Stream[F, Byte], bucket: Bucket, key: ObjectKey): F[Unit] =
+    Concurrent[F].bracketCase(
+      startUpload(client, bucket, key))(createResponse =>
+      for {
+        parts <- uploadParts(client, bytes, createResponse.uploadId, partSize, bucket, key)
+        _ <- completeUpload(client, createResponse.uploadId, bucket, key, parts)
+      } yield ()) {
+      case (createResponse, ExitCase.Canceled | ExitCase.Error(_)) =>
+        abortUpload(client, createResponse.uploadId, bucket, key).void
+      case (_, ExitCase.Completed) =>
+        Concurrent[F].unit
+    }
+
+  private def startUpload(client: S3AsyncClient, bucket: Bucket, key: ObjectKey)
+      : F[CreateMultipartUploadResponse] =
+    Concurrent[F].delay(client.createMultipartUpload(
+      CreateMultipartUploadRequest.builder.bucket(bucket.value).key(key.value).build)).futureLift
+
+  private def uploadParts(
+    client: S3AsyncClient,
+    bytes: Stream[F, Byte],
+    uploadId: String,
+    minChunkSize: Int,
+    bucket: Bucket,
+    key: ObjectKey): F[List[CompletedPart]] =
+    (bytes.chunkMin(minChunkSize).zipWithIndex evalMap {
+      case (byteChunk, n) => {
+        // parts numbers must start at 1
+        val partNumber = Int.box(n.toInt + 1)
+
+        val uploadPartRequest =
+          UploadPartRequest.builder
+            .bucket(bucket.value)
+            .uploadId(uploadId)
+            .key(key.value)
+            .partNumber(partNumber)
+            .contentLength(Long.box(byteChunk.size.toLong))
+            .build
+
+        val uploadPartResponse =
+          Concurrent[F].delay(
+            client.uploadPart(
+              uploadPartRequest,
+              AsyncRequestBody.fromByteBuffer(byteChunk.toByteBuffer))).futureLift
+
+        uploadPartResponse.map(response =>
+          CompletedPart.builder.partNumber(partNumber).eTag(response.eTag).build)
+      }
+    }).compile.toList
+
+  private def completeUpload(
+    client: S3AsyncClient,
+    uploadId: String,
+    bucket: Bucket,
+    key: ObjectKey,
+    parts: List[CompletedPart]): F[CompleteMultipartUploadResponse] = {
+    val multipartUpload =
+      CompletedMultipartUpload.builder.parts(parts :_*).build
```

Comment:
I generally prefer `parts: _*` here, to be uniform with conventional ascription styling.

Reply:
Ohh I didn't see splats as type ascription. But makes sense when you put it that way.